v1/ast/parser_test.go

```diff
@@ -6378,6 +6378,146 @@ allow if {
 	assertLocationText(t, "# METADATA\n# title: rule", m.Rules[0].Annotations[0].Location)
 }
 
+func TestMaxParsingRecursionDepth(t *testing.T) {
+	tests := []struct {
+		name        string
+		input       string
+		depthLimit  int
+		expectError bool
+	}{
+		{
+			name:        "deeply_nested_array_exceeds_limit",
+			input:       generateDeeplyNestedArray(1000),
+			depthLimit:  500,
+			expectError: true,
+		},
+		{
+			name:        "deeply_nested_array_within_limit",
+			input:       generateDeeplyNestedArray(100),
+			depthLimit:  500,
+			expectError: false,
+		},
+		{
+			name:        "deeply_nested_object_exceeds_limit",
+			input:       generateDeeplyNestedObject(1000),
+			depthLimit:  500,
+			expectError: true,
+		},
+		{
+			name:        "deeply_nested_object_within_limit",
+			input:       generateDeeplyNestedObject(100),
+			depthLimit:  500,
+			expectError: false,
+		},
+		{
+			name:        "deeply_nested_ref_exceeds_limit",
+			input:       generateDeeplyNestedRef(500),
+			depthLimit:  50,
+			expectError: true,
+		},
+		{
+			name:        "deeply_nested_ref_within_limit",
+			input:       generateDeeplyNestedRef(100),
+			depthLimit:  500,
+			expectError: false,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create module code
+			moduleCode := "package test\n\nvalue = " + tc.input
+
+			// Parse module
+			parser := NewParser().
+				WithFilename("test.rego").
+				WithReader(strings.NewReader(moduleCode)).
+				WithMaxRecursionDepth(tc.depthLimit)
+
+			_, _, errs := parser.Parse()
+
+			hasErr := len(errs) > 0
+			if hasErr != tc.expectError {
+				t.Errorf("Test %q: expected error: %v, got error: %v, error detail: %v", tc.name, tc.expectError, hasErr, errs)
+			}
+
+			if tc.expectError && hasErr {
+				// Verify the error contains our expected error
+				errFound := false
+				for _, e := range errs {
+					if strings.Contains(e.Message, ErrMaxParsingRecursionDepthExceeded.Error()) {
+						errFound = true
+						break
+					}
+				}
+				if !errFound {
+					t.Errorf("Expected error to contain %q, but got: %v",
+						ErrMaxParsingRecursionDepthExceeded.Error(), errs)
+				}
+			}
+		})
+	}
+}
+
+// generateDeeplyNestedArray creates a deeply nested array as a string
+// with the specified depth.
+func generateDeeplyNestedArray(depth int) string {
+	var sb strings.Builder
+
+	// Open brackets
+	for range depth {
+		sb.WriteString("[")
+	}
+
+	// Add a value at the deepest level
+	sb.WriteString("1")
+
+	// Close brackets
+	for range depth {
+		sb.WriteString("]")
+	}
+
+	return sb.String()
```

Comment:
Unless it's dog slow, we could probably just do something like:
```go
strings.Repeat("[", depth) + "1" + strings.Repeat("]", depth)
```

Reply:
Done, thanks!